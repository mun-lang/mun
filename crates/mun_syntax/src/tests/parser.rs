use crate::SourceFile;

fn ok_snapshot_test(text: &str) {
    let text = text.trim().replace("\n    ", "\n");
    let file = SourceFile::parse(&text);
    let errors = file.errors();
    assert_eq!(
        &*errors,
        &[] as &[crate::SyntaxError],
        "There should be no errors\nAST:\n{}",
        file.debug_dump()
    );
    insta::assert_snapshot!(insta::_macro_support::AutoName, file.debug_dump(), &text);
}

#[test]
fn empty() {
    ok_snapshot_test(r#""#);
}

#[test]
fn function() {
    ok_snapshot_test(
        r#"
    // Source file comment

    // Comment that belongs to the function
    fn a() {}
    fn b(value:number) {}
    export fn c():never {}
    fn b(value:number):number {}"#,
    );
}

#[test]
fn block() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a;
        let b:i32;
        let c:string;
    }"#,
    );
}

#[test]
fn literals() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a = true;
        let b = false;
        let c = 1;
        let d = 1.12;
        let e = "Hello, world!"
    }
    "#,
    );
}

#[test]
fn unary_expr() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a = --3;
        let b = !!true;
    }
    "#,
    )
}

#[test]
fn binary_expr() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a = 3+4*5
        let b = 3*4+10/2
    }
    "#,
    )
}

#[test]
fn expression_statement() {
    ok_snapshot_test(
        r#"
    fn foo() {
        let a = "hello"
        let b = "world"
        let c
        b = "Hello, world!"
        !-5+2*(a+b);
        -3
    }
    "#,
    )
}

#[test]
fn function_calls() {
    ok_snapshot_test(
        r#"
    fn bar(i:number) { }
    fn foo(i:number) {
      bar(i+1)
    }
    "#,
    )
}

#[test]
fn patterns() {
    ok_snapshot_test(
        r#"
    fn main(_:number) {
       let a = 0;
       let _ = a;
    }
    "#,
    )
}

#[test]
fn compare_operands() {
    ok_snapshot_test(
        r#"
    fn main() {
        let _ = a==b;
        let _ = a == b;
        let _ = a != b;
        let _ = a < b;
        let _ = a > b;
        let _ = a <= b;
        let _ = a >= b;
    }
    "#,
    )
}

#[test]
fn if_expr() {
    ok_snapshot_test(
        r#"
    fn bar() {
        if true {};
        if true {} else {};
        if true {} else if false {} else {};
        if {true} {} else {}
    }
    "#,
    );
}

#[test]
fn block_expr() {
    ok_snapshot_test(
        r#"
    fn bar() {
        {3}
    }
    "#,
    );
}

#[test]
fn return_expr() {
    ok_snapshot_test(
        r#"
    fn foo() {
        return;
        return 50;
    }
    "#,
    )
}

#[test]
fn loop_expr() {
    ok_snapshot_test(
        r#"
    fn foo() {
        loop {}
    }"#,
    )
}
