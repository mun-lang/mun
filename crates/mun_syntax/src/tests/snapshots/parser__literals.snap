---
source: crates/mun_syntax/src/tests/parser.rs
expression: "fn foo() {\n    let a = true;\n    let b = false;\n    let c = 1;\n    let d = 1.12;\n    let e = \"Hello, world!\"\n}"
---
SOURCE_FILE@[0; 110)
  FUNCTION_DEF@[0; 110)
    FN_KW@[0; 2) "fn"
    WHITESPACE@[2; 3) " "
    NAME@[3; 6)
      IDENT@[3; 6) "foo"
    PARAM_LIST@[6; 8)
      L_PAREN@[6; 7) "("
      R_PAREN@[7; 8) ")"
    WHITESPACE@[8; 9) " "
    BLOCK_EXPR@[9; 110)
      L_CURLY@[9; 10) "{"
      WHITESPACE@[10; 15) "\n    "
      LET_STMT@[15; 28)
        LET_KW@[15; 18) "let"
        WHITESPACE@[18; 19) " "
        BIND_PAT@[19; 20)
          NAME@[19; 20)
            IDENT@[19; 20) "a"
        WHITESPACE@[20; 21) " "
        EQ@[21; 22) "="
        WHITESPACE@[22; 23) " "
        LITERAL@[23; 27)
          TRUE_KW@[23; 27) "true"
        SEMI@[27; 28) ";"
      WHITESPACE@[28; 33) "\n    "
      LET_STMT@[33; 47)
        LET_KW@[33; 36) "let"
        WHITESPACE@[36; 37) " "
        BIND_PAT@[37; 38)
          NAME@[37; 38)
            IDENT@[37; 38) "b"
        WHITESPACE@[38; 39) " "
        EQ@[39; 40) "="
        WHITESPACE@[40; 41) " "
        LITERAL@[41; 46)
          FALSE_KW@[41; 46) "false"
        SEMI@[46; 47) ";"
      WHITESPACE@[47; 52) "\n    "
      LET_STMT@[52; 62)
        LET_KW@[52; 55) "let"
        WHITESPACE@[55; 56) " "
        BIND_PAT@[56; 57)
          NAME@[56; 57)
            IDENT@[56; 57) "c"
        WHITESPACE@[57; 58) " "
        EQ@[58; 59) "="
        WHITESPACE@[59; 60) " "
        LITERAL@[60; 61)
          INT_NUMBER@[60; 61) "1"
        SEMI@[61; 62) ";"
      WHITESPACE@[62; 67) "\n    "
      LET_STMT@[67; 80)
        LET_KW@[67; 70) "let"
        WHITESPACE@[70; 71) " "
        BIND_PAT@[71; 72)
          NAME@[71; 72)
            IDENT@[71; 72) "d"
        WHITESPACE@[72; 73) " "
        EQ@[73; 74) "="
        WHITESPACE@[74; 75) " "
        LITERAL@[75; 79)
          FLOAT_NUMBER@[75; 79) "1.12"
        SEMI@[79; 80) ";"
      WHITESPACE@[80; 85) "\n    "
      LET_STMT@[85; 108)
        LET_KW@[85; 88) "let"
        WHITESPACE@[88; 89) " "
        BIND_PAT@[89; 90)
          NAME@[89; 90)
            IDENT@[89; 90) "e"
        WHITESPACE@[90; 91) " "
        EQ@[91; 92) "="
        WHITESPACE@[92; 93) " "
        LITERAL@[93; 108)
          STRING@[93; 108) "\"Hello, world!\""
      WHITESPACE@[108; 109) "\n"
      R_CURLY@[109; 110) "}"

